### PR TITLE
feat: search filter expression

### PR DESCRIPTION
**Pull request by honeyAcorn**
This feature will filter the search results based on the expression user provided. We support `EQUALS`, `AND`, `OR`, `NOT`, `>`, `>=`, `<`, `<=` and `LIST_CONTAINS` for now.

### CODE REVIEW

File: proto/vectorindex.proto [hunk 1]
```diff
@@ -66,6 +66,81 @@ message _MetadataRequest {
...
+}
+
+message _NotExpression {
+    _FilterExpression not_expression = 1;
+}
+
+message _EqualsExpression {
+    string field = 1;
+    oneof value {
+        string string_val = 2;
```
Comment (malandis) [inline proto/vectorindex.proto:86]: style nit: to be consistent with the rest of the file, could you rename these to `*_value`? (see metadata above)

File: proto/vectorindex.proto [hunk 2]
```diff
@@ -66,6 +66,81 @@ message _MetadataRequest {
...
+}
+
+message _ListContainsExpression {
+    string field = 1;
+    string value = 2;
+}
+
+message _FilterExpression {
+    oneof expression {
+        _EqualsExpression equals_expr = 1;
```
Comment (malandis) [inline proto/vectorindex.proto:132]: style nit: `equals_expression` etc consistent with other names here

File: proto/vectorindex.proto [hunk 3]
```diff
@@ -112,25 +189,25 @@ message _SearchAndFetchVectorsResponse {
 }
 
 message _GetItemBatchRequest {
-  string index_name = 1;
-  repeated string ids = 2;
-  _MetadataRequest metadata_fields = 3;
+    string index_name = 1;
+    repeated string ids = 2;
+    _MetadataRequest metadata_fields = 3;
```
Comment (malandis) [inline proto/vectorindex.proto:194]: Did you run a code formatter or was this intended?

**Review by malandis — COMMENTED** [review]
Some minor style nits

File: proto/vectorindex.proto (hunk 1 above)
Comment (honeyAcorn) [reply proto/vectorindex.proto]: modified! https://github.com/momentohq/client-protos/pull/246/commits/0314a5fb4975ca699ac9f6fd2c8194e7d343375a

File: proto/vectorindex.proto (hunk 2 above)
Comment (honeyAcorn) [reply proto/vectorindex.proto]: modified! https://github.com/momentohq/client-protos/pull/246/commits/0314a5fb4975ca699ac9f6fd2c8194e7d343375a

File: proto/vectorindex.proto (hunk 3 above)
Comment (honeyAcorn) [reply proto/vectorindex.proto:194]: yes, if you check the above indentation, only these are wrong.

**Review by malandis — APPROVED** [review]

**Comment by honeyAcorn** [comment]
Please do not merge until backend is ready.

File: proto/vectorindex.proto [hunk 4]
```diff
@@ -66,6 +66,81 @@ message _MetadataRequest {
...
+    _FilterExpression second_expression = 2;
+}
+
+message _OrExpression {
+    _FilterExpression first_expression = 1;
+    _FilterExpression second_expression = 2;
+}
+
+message _NotExpression {
+    _FilterExpression not_expression = 1;
```
Comment (malandis) [inline proto/vectorindex.proto:80]:
```suggestion
    _FilterExpression expression_to_negate = 1;
```

**Review by malandis — COMMENTED** [review]
LGTM; just one naming nit

File: proto/vectorindex.proto (hunk 4 above)
Comment (honeyAcorn) [reply proto/vectorindex.proto:80]: Okay! Let me also fix this in backend first cause the two proto will have to sync.

Comment (honeyAcorn) [reply proto/vectorindex.proto]: Done: https://github.com/momentohq/client-protos/pull/246/commits/71fbd3c72bf5ff3d2f3567cf78cbdeebf929bf45

**Review by malandis — APPROVED** [review]

**Pull request merged**